Add <input type=checkbox switch> shadow tree

annevk opened the pull request:
#### 26e165bb2d52c3cd874938c8767c6b703b9ff949
<pre>
Add &lt;input type=checkbox switch&gt; shadow tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=263435">https://bugs.webkit.org/show_bug.cgi?id=263435</a>
rdar://117254677

Reviewed by Chris Dumez.

This conditionally adds a shadow tree with two elements when a checkbox
becomes a switch control. It also gives these two elements distinct
appearance values that depend on the shadow host&apos;s appearance value
being auto.

As checkboxes have been made performance-critical by forces outside of
my control, we add a bit to HTMLInputElement to track the presence of a
switch attribute, which needsShadowSubtree() can then rely on.

As we call removeShadowSubtree() while the InputType class stays alive,
we need to reset m_hasCreatedShadowSubtree to false as otherwise a
subsequent createShadowSubtreeIfNeeded() returns too early.

In preparation for future work we also ensure that RenderTheme does the
correct thing for isChecked, isEnabled, and isPressed, when it comes to
the new elements in the shadow tree.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::CheckboxInputType::createShadowSubtree):
(WebCore::CheckboxInputType::shouldAppearIndeterminate const):
(WebCore::CheckboxInputType::attributeChanged): Deleted.
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::attributeChanged):
(WebCore::HTMLInputElement::isSwitch const):
* Source/WebCore/html/HTMLInputElement.h:
(WebCore::HTMLInputElement::hasSwitchAttribute const):
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::removeShadowSubtree):
* Source/WebCore/html/InputType.h:
(WebCore::InputType::isSwitch const):
(WebCore::InputType::needsShadowSubtree const):
* Source/WebCore/html/shadow/SwitchThumbElement.cpp: Added.
(WebCore::SwitchThumbElement::create):
(WebCore::SwitchThumbElement::SwitchThumbElement):
(WebCore::SwitchThumbElement::resolveCustomStyle):
* Source/WebCore/html/shadow/SwitchThumbElement.h: Added.
* Source/WebCore/html/shadow/SwitchTrackElement.cpp: Added.
(WebCore::SwitchTrackElement::create):
(WebCore::SwitchTrackElement::SwitchTrackElement):
(WebCore::SwitchTrackElement::resolveCustomStyle):
* Source/WebCore/html/shadow/SwitchTrackElement.h: Added.
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autoAppearanceForElement const):
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::isChecked const):
(WebCore::RenderTheme::isEnabled const):
(WebCore::RenderTheme::isPressed const):
* Source/WebCore/style/StyleAppearance.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/style/StyleAppearance.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;ControlPart&gt;::encode):
(IPC::ArgumentCoder&lt;ControlPart&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/269869@main">https://commits.webkit.org/269869@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6fba5d7ec834879977853a9a534330b76a644527

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23883 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1995 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24978 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26029 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22014 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24157 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3604 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24379 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24125 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1538 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20640 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26620 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/1293 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21551 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/27792 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/21771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21831 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25581 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/1248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/18933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1256 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5707 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/1660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1577 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1284,7 +1285,9 @@ bool RenderTheme::isWindowActive(const RenderObject& renderer) const
-    return is<HTMLInputElement>(o.node()) && downcast<HTMLInputElement>(*o.node()).shouldAppearChecked();
+    if (is<HTMLInputElement>(o.node()))
+        return downcast<HTMLInputElement>(*o.node()).shouldAppearChecked();
+    return o.node()->shadowHost() && is<HTMLInputElement>(o.node()->shadowHost()) && downcast<HTMLInputElement>(*o.node()->shadowHost()).shouldAppearChecked();
```
Let's update to use `dynamicDowncast` here.

cdumez commented on Source/WebCore/html/shadow/SwitchThumbElement.cpp:
```diff
@@ -0,0 +1,61 @@
+
+Ref<SwitchThumbElement> SwitchThumbElement::create(Document& document)
+{
+    auto element = adoptRef(*new SwitchThumbElement(document));
```
`Ref element = ...`

cdumez commented on Source/WebCore/html/shadow/SwitchThumbElement.cpp:
```diff
@@ -0,0 +1,61 @@
+Ref<SwitchThumbElement> SwitchThumbElement::create(Document& document)
+{
+    auto element = adoptRef(*new SwitchThumbElement(document));
+    ScriptDisallowedScope::EventAllowedScope eventAllowedScope { element };
```
This goes out of scope right away, how is it useful? I don't get it.

annevk replied:
This was done in anticipation of future changes, but I've removed it for now.

cdumez commented on Source/WebCore/html/shadow/SwitchThumbElement.h:
```diff
@@ -0,0 +1,46 @@
+    static Ref<SwitchThumbElement> create(Document&);
+private:
+    constexpr static auto CreateSwitchThumbElement = CreateHTMLDivElement | NodeFlag::HasCustomStyleResolveCallbacks;
+    SwitchThumbElement(Document&);
```
explicit

cdumez commented on Source/WebCore/html/shadow/SwitchTrackElement.cpp:
```diff
@@ -0,0 +1,61 @@
+
+Ref<SwitchTrackElement> SwitchTrackElement::create(Document& document)
+{
+    auto element = adoptRef(*new SwitchTrackElement(document));
```
`Ref`

cdumez commented on Source/WebCore/html/shadow/SwitchTrackElement.cpp:
```diff
@@ -0,0 +1,61 @@
+Ref<SwitchTrackElement> SwitchTrackElement::create(Document& document)
+{
+    auto element = adoptRef(*new SwitchTrackElement(document));
+    ScriptDisallowedScope::EventAllowedScope eventAllowedScope { element };
```
Ditto, why is this useful?

cdumez commented on Source/WebCore/html/shadow/SwitchTrackElement.h:
```diff
@@ -0,0 +1,46 @@
+namespace WebCore {
+
+class HTMLInputElement;
+class TouchEvent;
```
Where is this used? I don't see it.

cdumez commented on Source/WebCore/html/shadow/SwitchTrackElement.h:
```diff
@@ -0,0 +1,46 @@
+    static Ref<SwitchTrackElement> create(Document&);
+private:
+    constexpr static auto CreateSwitchTrackElement = CreateHTMLDivElement | NodeFlag::HasCustomStyleResolveCallbacks;
+    SwitchTrackElement(Document&);
```
explicit

cdumez commented on Source/WebCore/html/shadow/SwitchTrackElement.h:
```diff
@@ -0,0 +1,46 @@
+public:
+    static Ref<SwitchTrackElement> create(Document&);
+private:
+    constexpr static auto CreateSwitchTrackElement = CreateHTMLDivElement | NodeFlag::HasCustomStyleResolveCallbacks;
```
Don't we usually use `static constexpr` in the code base instead of `constexpr static`? Same comment for the other class.

annevk replied:
We seem to use `static constexpr auto` outside of DOM more, but in DOM this is used more. Happy to change this here (and elsewhere in a separate PR).

cdumez commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1284,7 +1285,11 @@ bool RenderTheme::isWindowActive(const RenderObject& renderer) const
-    return is<HTMLInputElement>(o.node()) && downcast<HTMLInputElement>(*o.node()).shouldAppearChecked();
+    if (auto* element = dynamicDowncast<HTMLInputElement>(o.node()))
+        return element->shouldAppearChecked();
+    if (auto* host = dynamicDowncast<HTMLInputElement>(o.node()->shadowHost()))
```
What if `o.node()` returns nullptr?

The previous code was dealing with that fine. the new code would crash.

cdumez commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1284,7 +1285,13 @@ bool RenderTheme::isWindowActive(const RenderObject& renderer) const
-    return is<HTMLInputElement>(o.node()) && downcast<HTMLInputElement>(*o.node()).shouldAppearChecked();
+    if (!o.node())
+        return false;
+    if (auto* element = dynamicDowncast<HTMLInputElement>(o.node()))
```
`*o.node()` now that you null check above.

cdumez reviewed (approved):
LGTM with minor fix.

annevk committed:
Add <input type=checkbox switch> shadow tree
https://bugs.webkit.org/show_bug.cgi?id=263435
rdar://117254677

Reviewed by Chris Dumez.

This conditionally adds a shadow tree with two elements when a checkbox
becomes a switch control. It also gives these two elements distinct
appearance values that depend on the shadow host's appearance value
being auto.

As checkboxes have been made performance-critical by forces outside of
my control, we add a bit to HTMLInputElement to track the presence of a
switch attribute, which needsShadowSubtree() can then rely on.

As we call removeShadowSubtree() while the InputType class stays alive,
we need to reset m_hasCreatedShadowSubtree to false as otherwise a
subsequent createShadowSubtreeIfNeeded() returns too early.

In preparation for future work we also ensure that RenderTheme does the
correct thing for isChecked, isEnabled, and isPressed, when it comes to
the new elements in the shadow tree.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::CheckboxInputType::createShadowSubtree):
(WebCore::CheckboxInputType::shouldAppearIndeterminate const):
(WebCore::CheckboxInputType::attributeChanged): Deleted.
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::attributeChanged):
(WebCore::HTMLInputElement::isSwitch const):
* Source/WebCore/html/HTMLInputElement.h:
(WebCore::HTMLInputElement::hasSwitchAttribute const):
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::removeShadowSubtree):
* Source/WebCore/html/InputType.h:
(WebCore::InputType::isSwitch const):
(WebCore::InputType::needsShadowSubtree const):
* Source/WebCore/html/shadow/SwitchThumbElement.cpp: Added.
(WebCore::SwitchThumbElement::create):
(WebCore::SwitchThumbElement::SwitchThumbElement):
(WebCore::SwitchThumbElement::resolveCustomStyle):
* Source/WebCore/html/shadow/SwitchThumbElement.h: Added.
* Source/WebCore/html/shadow/SwitchTrackElement.cpp: Added.
(WebCore::SwitchTrackElement::create):
(WebCore::SwitchTrackElement::SwitchTrackElement):
(WebCore::SwitchTrackElement::resolveCustomStyle):
* Source/WebCore/html/shadow/SwitchTrackElement.h: Added.
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autoAppearanceForElement const):
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::isChecked const):
(WebCore::RenderTheme::isEnabled const):
(WebCore::RenderTheme::isPressed const):
* Source/WebCore/style/StyleAppearance.cpp:
(WebCore::operator<<):
* Source/WebCore/style/StyleAppearance.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<ControlPart>::encode):
(IPC::ArgumentCoder<ControlPart>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/269869@main
